src/NuGet.Clients/PackageManagement.VisualStudio/Migrate/ProjectJsonToPackageRefMigrator.cs

```diff
@@ -87,49 +100,55 @@ private static void MigrateRuntimes(PackageSpec packageSpec, Microsoft.Build.Eva
             if (runtimes != null && runtimes.Count > 0)
             {
                 runtimeIdentifiers.AddRange(runtimes.Keys);
-                
+
             }
 
-            if(supports != null && supports.Count > 0)
+            if (supports != null && supports.Count > 0)
             {
                 runtimeSupports.AddRange(supports.Keys);
             }
 
             var union = string.Join(";", runtimeIdentifiers.Union(runtimeSupports));
-            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            await NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async delegate
             {
                 await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
                 buildProject.SetProperty("RuntimeIdentifiers", union);
-            });            
+            });
+
         }
 
-        private static void RemoveProjectJsonReference(Microsoft.Build.Evaluation.Project buildProject, string projectJsonFilePath)
+        private static async Task RemoveProjectJsonReferenceAsync(Microsoft.Build.Evaluation.Project buildProject, string projectJsonFilePath)
         {
-            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
-            {
-                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                var projectJsonItem = buildProject.GetItems("None")
+            var projectJsonItem = buildProject.GetItems("None")
```

Comment:
You should ensure as an async method that you're on the UI thread before accessing MSBuild, if this isn't a CPS project.
I see you switch to it *later* in this method. So I suggest you just move that up to the top.
Hint: the rule applies anytime you use MSBuild at all -- not just for writing.